pycc/ccresponse.py

```diff
@@ -0,0 +1,361 @@
+                print("\nPerturbed wave function converged in %.3f seconds.\n" % (time.time() - solver_start))
+                return self.X1, self.X2, pseudo
+
+            diis.add_error_vector(self.X1, self.X2)
+            if niter >= start_diis:
+                self.X1, self.X2 = diis.extrapolate(self.X1, self.X2)
+
+    def r_X1(self, pertbar, omega):
+        contract = self.contract
+        o = self.ccwfn.o
+        v = self.ccwfn.v
+        X1 = self.X1
+        X2 = self.X2
+        hbar = self.hbar
+
+        r_X1 = (pertbar.Avo.T - omega * X1).copy()
+        r_X1 += contract('ie,ae->ia', X1, hbar.Hvv)
+        r_X1 -= contract('ma,mi->ia', X1, hbar.Hoo)
+        r_X1 += 2.0*contract('me,maei->ia', X1, hbar.Hovvo)
+        r_X1 -= contract('me,maie->ia', X1, hbar.Hovov)
+        r_X1 += contract('me,miea->ia', hbar.Hov, (2.0*X2 - X2.swapaxes(0,1)))
+        r_X1 += contract('imef,amef->ia', X2, (2.0*hbar.Hvovv - hbar.Hvovv.swapaxes(2,3)))
+        r_X1 -= contract('mnae,mnie->ia', X2, (2.0*hbar.Hooov - hbar.Hooov.swapaxes(0,1)))
+
+        return r_X1
+
+    def r_X2(self, pertbar, omega):
+        contract = self.contract
+        o = self.ccwfn.o
+        v = self.ccwfn.v
+        X1 = self.X1
+        X2 = self.X2
+        t2 = self.ccwfn.t2
+        hbar = self.hbar
+        L = self.H.L
+
+        Zvv = contract('amef,mf->ae', (2.0*hbar.Hvovv - hbar.Hvovv.swapaxes(2,3)), X1)
+        Zvv -= contract('mnef,mnaf->ae', L[o,o,v,v], X2)
+
+        Zoo = -1.0*contract('mnie,ne->mi', (2.0*hbar.Hooov - hbar.Hooov.swapaxes(0,1)), X1)
+        Zoo -= contract('mnef,inef->mi', L[o,o,v,v], X2)
+
+        r_X2 = pertbar.Avvoo - 0.5 * omega*X2
+        r_X2 += contract('ie,abej->ijab', X1, hbar.Hvvvo)
+        r_X2 -= contract('ma,mbij->ijab', X1, hbar.Hovoo)
+        r_X2 += contract('mi,mjab->ijab', Zoo, t2)
+        r_X2 += contract('ae,ijeb->ijab', Zvv, t2)
+        r_X2 += contract('ijeb,ae->ijab', X2, hbar.Hvv)
+        r_X2 -= contract('mjab,mi->ijab', X2, hbar.Hoo)
+        r_X2 += 0.5*contract('mnab,mnij->ijab', X2, hbar.Hoooo)
+        r_X2 += 0.5*contract('ijef,abef->ijab', X2, hbar.Hvvvv)
+        r_X2 -= contract('imeb,maje->ijab', X2, hbar.Hovov)
+        r_X2 -= contract('imea,mbej->ijab', X2, hbar.Hovvo)
+        r_X2 += 2.0*contract('miea,mbej->ijab', X2, hbar.Hovvo)
+        r_X2 -= contract('miea,mbje->ijab', X2, hbar.Hovov)
+
+        r_X2 = r_X2 + r_X2.swapaxes(0,1).swapaxes(2,3)
+
+        return r_X2
+
+    def pseudoresponse(self, pertbar, X1, X2):
+        contract = self.ccwfn.contract
+        polar1 = 2.0 * contract('ai,ia->', np.conj(pertbar.Avo), X1)
+        polar2 = 2.0 * contract('ijab,ijab->', np.conj(pertbar.Avvoo), (2.0*X2 - X2.swapaxes(2,3)))
+
+        return -2.0*(polar1 + polar2)
+        
+
+class pertbar(object):
+    def __init__(self, pert, ccwfn):
+        o = ccwfn.o
+        v = ccwfn.v
+        t1 = ccwfn.t1
+        t2 = ccwfn.t2
+        contract = opt_einsum.contract
```

Comment:
could you just `contract = ccwfn.contract` as in the initialization of the `ccresponse` object? this would 1) remove the need for `import opt_einsum` at the top, and 2) allow GPU support (I think all we would have to do is add `if` statements to tensor initializations, @zw0930 ?)